en/lessons/libraries/bypass.md

```diff
@@ -0,0 +1,274 @@
+
+With our code in place we need to start on our tests.
+Before we can use Bypass we'll need to ensure it's running.
+To do that, let's update `test/test_helper.exs` look like this:
+
+```elixir
+ExUnit.start()
+Application.ensure_all_started(:bypass)
+```
+
+Now that we know Bypass will be running during our tests let's head over to `test/clinic/health_check_test.exs` and finish our setup.
+To prepare Bypass for accepting requests we need to open the connect with `Bypass.open/1`, this can be done in our test setup callback:
+
+```elixir
+defmodule Clinic.HealthCheckTests do
+  use ExUnit.Case
+
+  setup do
+    bypass = Bypass.open()
+    {:ok, bypass: bypass}
+  end
+end
+```
+
+We're ready to put Bypass to work!  We'll start with a successful request first:
+
+```elixir
+defmodule Clinic.HealthCheckTests do
+  use ExUnit.Case
+
+  alias Clinic.HealthCheck
+
+  setup do
+    bypass = Bypass.open()
+    {:ok, bypass: bypass}
+  end
+
+  test "request with HTTP 200 response", %{bypass: bypass} do
+    Bypass.expect(bypass, fn conn ->
+      Plug.Conn.resp(conn, 200, "pong")
+    end)
+
+    assert {:ok, "pong"} = HealthCheck.ping("http://localhost:#{bypass.port}")
+  end
+end
+```
+
+Our test is simple enough and if we run it we'll see it passes but let's dig in and see what each portion is doing.
+The first thing we see in our test is the `Bypass.expect/2` function:
+
+```elixir
+Bypass.expect(bypass, fn conn ->
+  Plug.Conn.resp(conn, 200, "pong")
+end)
+
+```
+
+`Bypass.expect/2` takes our Bypass connection and a single arity function which is expected to modify a connection and return it, this is also an opportunity to make assertions on the request to verify it's as we expect.  Let's update our test url to include `/ping` and assert both the request path and HTTP method:
+
+```elixir
+test "request with HTTP 200 response", %{bypass: bypass} do
+  Bypass.expect(bypass, fn conn ->
+    assert "GET" == conn.method
+    assert "/ping" == conn.request_path
+    Plug.Conn.resp(conn, 200, "pong")
+  end)
+
+  assert {:ok, "pong"} = HealthCheck.ping("http://localhost:#{bypass.port}/ping")
+end
+```
+
+The last part of our test we use `HealthCheck.ping/1` and assert the response is as expected, but what's `bypass.port` all about?
+Bypass is actually listening to a local port and intercepting those requests, we're using `bypass.port` to retrieve the default port since we didn't provide on in `Bypass.open/1`.
+
+Next up is adding test cases for errors.
+We can start with a test much like our first with some minor changes: returning 500 as the status code and assert the `{:error, reason}` tuple is returned:
+
+```elixir
+test "request with HTTP 500 response", %{bypass: bypass} do
+  Bypass.expect(bypass, fn conn ->
+    Plug.Conn.resp(conn, 500, "Server Error")
+  end)
+
+  assert {:error, "HTTP Status 500"} = HealthCheck.ping("http://localhost:#{bypass.port}")
```

Comment:
`assert {:error, "HTTP status 500"} ...`
lower s on status :)

Reply:
The `HealthCheck` module actually returns the string `HTTP Status #{status_code}".

Reply:
Oops ><